### PR TITLE
DrivAerML: FiLM conditioning on global geometry statistics

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DM Transolver processes each surface point token through the transformer with only local context (Fourier-encoded coordinates + attention). It has **no explicit representation of global car geometry** — the model must implicitly infer whether it's looking at a sedan, SUV, or fastback from the distribution of point positions alone. We hypothesize that explicitly providing global geometry statistics as conditioning signals will help the model adapt its per-point predictions to the specific car shape.

**FiLM conditioning** (Feature-wise Linear Modulation, Perez et al. 2018, arXiv:1709.07871) modulates hidden representations by applying a learned affine transform conditioned on external context: `h' = gamma(c) * h + beta(c)`, where c is a global context vector and gamma/beta are learned projections. This is:
1. **Standard in conditional generation** — used in StyleGAN, conditional diffusion models, and conditional neural fields
2. **Physics-informed for CFD** — global geometry statistics (bounding box dimensions, aspect ratio, frontal area proxy) directly affect aerodynamic behavior (drag, pressure distribution, separation patterns)
3. **Computationally cheap** — adds only a small MLP + per-layer gamma/beta projections
4. **Complementary to existing features** — Fourier encoding captures local position; FiLM captures global shape context

The intuition: a point at coordinates (x=0.3, y=0.1, z=0.5) on a long sedan has very different flow physics than the same coordinates on a short hatchback. FiLM conditioning gives the model this distinction explicitly.

### Implementation

1. **Add CLI flag:**
   ```python
   parser.add_argument('--geometry-film', action='store_true', default=False,
                       help='Add FiLM conditioning on global geometry statistics')
   ```

2. **Compute global geometry features** from the surface point cloud (in the forward pass, before transformer layers):
   ```python
   if args.geometry_film:
       # coords: [B, N, 3] — the raw xyz coordinates of all surface points
       geo_min = coords.min(dim=1).values  # [B, 3] — bounding box min
       geo_max = coords.max(dim=1).values  # [B, 3] — bounding box max
       geo_range = geo_max - geo_min        # [B, 3] — bounding box dimensions
       geo_center = coords.mean(dim=1)      # [B, 3] — centroid
       geo_std = coords.std(dim=1)          # [B, 3] — spread per axis
       
       # Concatenate into a global geometry vector
       geo_features = torch.cat([geo_min, geo_max, geo_range, geo_center, geo_std], dim=-1)  # [B, 15]
   ```

3. **Create FiLM conditioning module:**
   ```python
   class FiLMLayer(nn.Module):
       def __init__(self, geo_dim, hidden_dim):
           super().__init__()
           self.film_mlp = nn.Sequential(
               nn.Linear(geo_dim, hidden_dim),
               nn.GELU(),
               nn.Linear(hidden_dim, hidden_dim * 2)  # gamma and beta
           )
       
       def forward(self, h, geo_features):
           # h: [B, N, D], geo_features: [B, geo_dim]
           film_params = self.film_mlp(geo_features)  # [B, 2*D]
           gamma, beta = film_params.chunk(2, dim=-1)  # [B, D] each
           gamma = gamma.unsqueeze(1)  # [B, 1, D]
           beta = beta.unsqueeze(1)    # [B, 1, D]
           return gamma * h + beta
   ```

4. **Add one FiLM layer per transformer layer.** Apply after each transformer layer's output (post-residual, post-LayerNorm):
   ```python
   # In model __init__:
   if args.geometry_film:
       self.film_layers = nn.ModuleList([
           FiLMLayer(geo_dim=15, hidden_dim=hidden_dim) 
           for _ in range(num_layers)
       ])
   
   # In forward, after each transformer layer i:
   if hasattr(self, 'film_layers'):
       x = self.film_layers[i](x, geo_features)
   ```

### Trials

**Smoke test first (3 epochs)** — verify shapes, loss finite, FiLM params non-degenerate (gamma not collapsing to 1, beta not collapsing to 0).

**Trial 1 — FiLM conditioning, standard config:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --geometry-film \
  --wandb_group dm-geometry-film
```

**Trial 2 — FiLM + additive only (no gamma, just beta — simpler conditioning):**
Modify the FiLM layer to output only beta (no multiplicative gamma), to test whether additive conditioning is sufficient:
```python
# In FiLMLayer forward:
beta = self.film_mlp(geo_features).unsqueeze(1)  # [B, 1, D]
return h + beta  # additive only
```
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --geometry-film --geometry-film-additive-only \
  --wandb_group dm-geometry-film
```

Log the gamma and beta statistics (mean, std) at each layer every 50 steps — this tells us whether the FiLM is actually learning geometry-dependent modulation or collapsing to identity.

Report `val_primary/surface_rel_l2_pct` for both trials. Target: beat **3.833%**.

## Baseline

Current best DrivAerML metrics (PR #3072, eren/dm-ema-9995-gc05):

| Metric | Value | Epoch |
|--------|-------|-------|
| val_primary/surface_rel_l2_pct | **3.833%** | 511 |
| test_primary/surface_rel_l2_pct | **4.685%** | 511 |

- W&B run: `ncl1dh88`
- Config: 4L/512d/8H, AdamW lr=5e-4, T_max=30, gc=0.5, EMA=0.9995, no WD, Fourier

Reproduce:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT brook:

## Results

### Trial 1 — Full FiLM conditioning (gamma + beta)

| Metric | Value | Epoch |
|--------|-------|-------|
| val_primary/surface_rel_l2_pct | **5.221%** | 135 |
| test_primary/surface_rel_l2_pct | **6.079%** | 135 |

- **Baseline val:** 3.833% — **not beaten** (36% worse)
- **Baseline test:** 4.685% — **not beaten** (30% worse)
- W&B run: `n68w02b8`
- Peak memory: ~6.5 GB (H100)
- **Training diverged at epoch 150** — gradient norms exploded from ~0.5 to 53k+, loss went to NaN. Best checkpoint preserved from epoch 135.

Command:
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --geometry-film \
  --wandb-group dm-geometry-film \
  --agent brook --wandb-name "brook/dm-film-full"
```

### Trial 2 — Additive-only FiLM (beta only, no gamma)

| Metric | Value | Epoch |
|--------|-------|-------|
| val_primary/surface_rel_l2_pct | **5.369%** | 174 |
| test_primary/surface_rel_l2_pct | **6.191%** | 174 |

- **Baseline val:** 3.833% — **not beaten** (40% worse)
- **Baseline test:** 4.685% — **not beaten** (32% worse)
- W&B run: `jbahi7mg`
- Peak memory: ~6.1 GB (H100)
- **Training diverged at epoch 178** — same instability pattern as Trial 1, but ~28 epochs later. Briefly recovered (epochs 184-192) before permanently diverging at epoch 193.

Command:
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --geometry-film --geometry-film-additive-only \
  --wandb-group dm-geometry-film \
  --agent brook --wandb-name "brook/dm-film-additive-only"
```

### What happened

**Neither trial beats the baseline.** Both FiLM variants suffer from training instability that leads to gradient explosion and permanent divergence:

- **Full FiLM (gamma+beta)** reached 5.22% val before diverging at epoch 150. The multiplicative gamma component amplifies gradients over time — grad norms went from 0.5 → 53,000+ within 15 epochs.
- **Additive-only (beta)** was more stable, reaching 5.37% val and lasting 28 epochs longer (diverged at epoch 178). It briefly self-recovered (epochs 184-192) before permanently diverging.

The core issue: FiLM conditioning injects an extra pathway that, combined with the cosine LR restart schedule (T_max=30), creates periodic gradient spikes that eventually cascade into divergence. Before diverging, both variants were converging well (~0.1% improvement per 10 epochs), suggesting the hypothesis has merit but the implementation needs stability fixes.

The hypothesis that global geometry context helps is **partially supported** — both variants converged significantly during early training. But the conditioning mechanism introduces instability that prevents them from reaching the baseline's 500+ epoch training horizon.

### Suggested follow-ups

1. **Gradient clipping on FiLM parameters specifically** — the divergence originates from the FiLM MLP weights growing large. A separate, tighter grad clip on FiLM params (e.g., 0.1) could prevent the cascade.
2. **Clamping delta_gamma to [-1, 1]** — prevents the multiplicative factor from exceeding [0, 2], which the DiT literature recommends for stability.
3. **Smaller FiLM MLP hidden dim** — using hidden_dim=64 instead of 512 would reduce the conditioning capacity and might prevent overfitting to geometry stats.
4. **Normalize geometry features offline** — compute StandardScaler stats over the training set and apply fixed normalization rather than relying on per-batch LayerNorm.
5. **Longer cosine period for FiLM** — the T_max=30 LR restarts create the sharpest gradient spikes. A longer T_max or LR warmup after restart could help.
6. **adaLN-Zero placement** — applying FiLM as the LayerNorm affine params (replacing LN's learned affine) rather than post-block, as in DiT. This requires modifying the architecture files but is the canonical stable approach.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3270 because ADVISOR: Closing as dead end. Both FiLM variants fail decisively:
- Full FiLM (gamma+beta): 5.221% val — 36% worse than baseline 3.833%, diverged at epoch 150 with grad norms 0.5→53,000+
- Additive-only (beta): 5.369% val — 40% worse, diverged at epoch 178

The gradient explosion is severe and the pre-divergence performance suggests the FiLM conditioning pathway fundamentally interferes with the Transolver's existing routing mechanism rather than complementing it.

The student's suggestion of adaLN-Zero (DiT-style conditioning within LayerNorm) is the more principled approach, but given our time constraints under #3283 we need to focus GPU cycles on the prioritized breakout lanes (AB-UPT anchored decoder, coord normalization, metric-aware fine-tune, domain-normalized volume). GPU reassigned to higher-priority work..
